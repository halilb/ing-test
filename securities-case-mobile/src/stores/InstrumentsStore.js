import { action, computed, observable, runInAction, useStrict } from 'mobx';
import Instrument from '../models/Instrument';

// allow mutations only in @action blocks
useStrict(true);

export default class InstrumentsStore {
  /*
   * The helper reference that'll manage API requests for us
   */
  transportLayer = null;
  /*
   * main list that holds Instrument instances
   */
  @observable instruments = null;

  /*
   * this is used to calculate priceFetchDiffSeconds
   */
  @observable lastPriceFetchTime = null;

  constructor(transportLayer, lastPriceFetchTime) {
    this.transportLayer = transportLayer;
    this.instruments = [];
    this.lastPriceFetchTime = lastPriceFetchTime;

    this.loadInstruments();
    this.transportLayer.initializeSocket(this.updateInstrumentPrice);
  }

  // moves amount and when properties to the top level
  parseInstrumentData = data => {
    const { code, price } = data;
    const { amount, when } = price;
    return { code, amount, when };
  };

  @action updateInstrumentPrice = data => {
    const { code, amount, when } = this.parseInstrumentData(data);
    const instrument = this.instruments.find(i => i.code === data.code);

    if (instrument) {
      instrument.price = amount;
      instrument.lastPriceTime = when;
      this.lastPriceFetchTime = Date.now();
    }
  };

  @action loadInstruments = async () => {
    const data = await this.transportLayer.fetchInstruments();
    const newInstruments = data.map(d => {
      const { code, amount, when } = this.parseInstrumentData(d);
      return new Instrument(code, amount, when);
    });

    runInAction('update instrument list after fetching network data', () => {
      this.instruments.replace(newInstruments);
      this.lastPriceFetchTime = Date.now();
    });

    return newInstruments;
  };

  // using dataSource to normalize observable array
  @computed get dataSource() {
    // updating an instrument instance price does not invoke this block weirdly.
    // so, I'm adding lastPriceFetchTime reference here as an ugly hack to invoke this.
    // one other solution is to recreate Instrument instance when price changes,
    // and I think that's even uglier.
    // I hope to change this when I find a better way.
    console.log(this.lastPriceFetchTime);
    return this.instruments.slice();
  }

  // used to display an information message whether the prices are up to date or not
  @computed get priceFetchDiffSeconds() {
    return (this.lastPriceFetchTime - Date.now()) / 1000;
  }
}
