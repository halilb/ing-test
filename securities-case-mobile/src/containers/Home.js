import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Text, View, StyleSheet } from 'react-native';
import { observer } from 'mobx-react/native';

import InstrumentsStore from '../stores/InstrumentsStore';
import { InstrumentList } from '../components';
import { commonStyles } from '../helpers/theme';

@observer
export default class Home extends Component {
  static propTypes = {
    instrumentsStore: PropTypes.instanceOf(InstrumentsStore),
  };

  render() {
    const {
      dataSource,
      priceFetchDiffSeconds,
    } = this.props.instrumentsStore;

    let priceTimeInfo;
    if (priceFetchDiffSeconds > 10) {
      priceTimeInfo = `Prices fetched ${priceFetchDiffSeconds} seconds ago`;
    } else {
      priceTimeInfo = 'Prices are up to date!';
    }

    return (
      <View style={styles.container}>
        <Text style={styles.priceTimeInfo}>{priceTimeInfo}</Text>
        <InstrumentList style={commonStyles.card} instruments={dataSource} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 24,
    paddingTop: 8,
  },
  priceTimeInfo: {
    textAlign: 'center',
    marginBottom: 8,
  },
});
