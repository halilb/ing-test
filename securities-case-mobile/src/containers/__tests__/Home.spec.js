import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import InstrumentsStore from '../../stores/InstrumentsStore';
import Home from '../Home';

// dummy transport layer object
const transportLayer = {
  initializeSocket: () => {},
};

it('renders a Home container correctly', () => {
  const fakeTime = 1497888703132;
  // mocking Date.now to make priceFetchDiffSeconds work
  Date.now = jest.fn(() => fakeTime);

  const store = new InstrumentsStore(transportLayer, fakeTime);
  const snapshot = renderer.create(<Home instrumentsStore={store} />).toJSON();

  expect(snapshot).toMatchSnapshot();
});

it('displays latest price fetch time correctly if it is older than 10 seconds', () => {
  const fakeTime = 1497888703132;
  // fakeTime is 11 seconds later than Date.now
  Date.now = jest.fn(() => fakeTime - 11 * 1000);

  const store = new InstrumentsStore(transportLayer, fakeTime);
  const snapshot = renderer.create(<Home instrumentsStore={store} />).toJSON();

  expect(snapshot).toMatchSnapshot();
});
