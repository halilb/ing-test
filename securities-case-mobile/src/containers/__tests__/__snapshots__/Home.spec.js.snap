// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays latest price fetch time correctly if it is older than 10 seconds 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 24,
      "paddingTop": 8,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "marginBottom": 8,
        "textAlign": "center",
      }
    }
  >
    Prices fetched 11 seconds ago
  </Text>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#ffffff",
        "borderRadius": 2,
        "borderWidth": 1,
        "padding": 8,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.71,
        "shadowRadius": 2,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`renders a Home container correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 24,
      "paddingTop": 8,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "marginBottom": 8,
        "textAlign": "center",
      }
    }
  >
    Prices are up to date!
  </Text>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#ffffff",
        "borderRadius": 2,
        "borderWidth": 1,
        "padding": 8,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.71,
        "shadowRadius": 2,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
